### ISSUE TITLE
Override Time Schedule

### ISSUE THREAD

**Issue by sjmcculloch** [NSWSESMembers/availability-poc#122]
The requirements state that the time schedule can be overwritten at any time:

_[1.30] At any time, a pattern of availability may be overridden for a single day without breaking the ongoing pattern._

e.g. I'm available Mon-Fri 9a-5p except this Wednesday I have an event on and now I want to mark myself unavailable for it. 

How does this happen in the existing model?

**Comment by sdunster** [comment]
Model currently doesn't really support this notion. My plan is to make it support any kind of time entry first with no support for "patterns". Feel free to suggest an evolution of the model that supports this, though if it's something you need in there to become unblocked. If you don't I'll do it when I get to it.

The way I see it the models evolves based on the features we need, not the other way around but if you want to work on building something exhaustive so you don't have to think about it later that's fine.

**Comment by sjmcculloch** [comment]
Patterns can be tricky so I can see why you went with the simpler approach to begin with.

We'll need to set expectations on what might be in a beta, so once I clean up the requirements for the web app as well, it would be a good idea that we sit down and work out what parts we can realistically build for a first release (and what features are stretch goals). 

For data model on patterns (I do have some experience) and it really depends on the types you want to store. 

This is a simple pattern:

Repeat this event every Tuesday & Thursday at times of 5p-10p.

The tables might look like:
Schedule
- ID
- Name
- Details
- StartTime
- EndTime

ScheduleRepeat
- RepeatID
- ScheduleID
- MetaKey
- MetaValue

In the repeat you store a series of "repeat_start" with no time (unix timestamp) and what the interval is in seconds.

repeat_interval_1 goes with repeat_start of the ID 1. If an event repeats every Tuesday and every Thursday, the repeat_interval would be 604800 (7 days), and there would be 2 repeat_starts and 2 repeat_intervals. 

RepeatID   ScheduleID      MetaKey                  MetaValue
1                  1                       repeat_start            1298959200
2                 1                        repeat_interval_1    604800
3                  1                       repeat_start            1299132000
4                 1                        repeat_interval_3    604800

And here is what the SQL looks like for a given day:

SELECT EV.*
FROM `events` EV
RIGHT JOIN `events_meta` EM1 ON EM1.`event_id` = EV.`id`
RIGHT JOIN `events_meta` EM2 ON EM2.`meta_key` = CONCAT( 'repeat_interval_', EM1.`id` )
WHERE EM1.meta_key = 'repeat_start'
    AND (
        ( CASE ( 1299132000 - EM1.`meta_value` )
            WHEN 0
              THEN 1
            ELSE ( 1299132000 - EM1.`meta_value` )
          END
        ) / EM2.`meta_value`
    ) = 1
LIMIT 0 , 30

It's an extension of this idea for more complex patterns:

Event A repeats every month on the 3rd of the month starting on July 1, 2017

**Comment by sdunster** [comment]
seems legit. I'm confused why you wouldn't denormalise into a table like this though:
schedule_id: ID of schedule ie 1
repeat_id: unique ID of repeat record for a schedule ie 1
starttime: timestamp of reference start time
endtime: timestamp of end time
repeat_interval: seconds between each schedule

your example would then be represented by 2 rows (same col order):
1 1 1298959200 1298959200+(60*60*5) 604800 // tuesday 5p-10p
1 2 1298959200 1299132000+(60*60*5) 604800 // thursday 5p-10p

I like the idea though sounds great and go ahead if you want to build it. I'm going to avoid it for now though for the reasons I said before.

Something worth keeping in mind is that this solution will get screwed by daylight savings changes because we're working entirely with timestamps and time intervals rather than calendar-based intervals.

**Comment by sjmcculloch** [comment]
Closing - no longer relevant.
